### PR TITLE
8266293: Key protection using PBEWithMD5AndDES fails with "java.security.InvalidAlgorithmParameterException: Salt must be 8 bytes long"

### PR DESCRIPTION
**Pull request by wangweij**
`PKCS12KeyStore` always uses a 20-byte salt in encryption but PBEWithMD5AndDES only accepts 8-byte salt. With this code change, the salt used for this algorithm will be 8 bytes.

RFC 2898 only requires the salt to be at least 8 bytes, but I don't intend to modify the `PBES1Core.java` to accept a long salt. Otherwise, a newly generated PKCS #⁠12 file using a long salt will not be recognized by an old JDK.

Also, although `PBES1Core.java` also take cares of another algorithm named PBEWithMD5AndDESede but it's not usable in a PKCS #⁠12 keystore as we have not defined its Object Identifier anywhere.

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8266293](https://bugs.openjdk.java.net/browse/JDK-8266293): Key protection using PBEWithMD5AndDES fails with "java.security.InvalidAlgorithmParameterException: Salt must be 8 bytes long"

### Reviewers
 * [Valerie Peng](https://openjdk.java.net/census#valeriep) (@valeriepeng - **Reviewer**) ⚠️ Review applies to 6726d0be77396f2ff79065b6822b5c306fb77d68

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/3822/head:pull/3822` \
`$ git checkout pull/3822`

Update a local copy of the PR: \
`$ git checkout pull/3822` \
`$ git pull https://git.openjdk.java.net/jdk pull/3822/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 3822`

View PR using the GUI difftool: \
`$ git pr show -t 3822`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/3822.diff">https://git.openjdk.java.net/jdk/pull/3822.diff</a>

</details>

### CODE REVIEW

**Comment by valeriepeng** [comment]
I will take a look. Thanks.

File: src/java.base/share/classes/sun/security/pkcs12/PKCS12KeyStore.java [hunk 1]
```diff
@@ -804,11 +804,17 @@ public synchronized void engineSetKeyEntry(String alias, byte[] key,
      */
     private AlgorithmParameters getPBEAlgorithmParameters(
             String algorithm, int iterationCount) throws IOException {
-        AlgorithmParameters algParams = null;
+        AlgorithmParameters algParams;
+
+        byte[] salt = getSalt();
+        if (KnownOIDs.findMatch(algorithm) == KnownOIDs.PBEWithMD5AndDES) {
+            // PBEWithMD5AndDES requires a 8-byte salt
```
Comment (valeriepeng) [inline src/java.base/share/classes/sun/security/pkcs12/PKCS12KeyStore.java:811]: nit: maybe use "PBES1 scheme such as PBEWithMD5AndDES requires a 8-byte salt"

Comment (wangweij) [reply src/java.base/share/classes/sun/security/pkcs12/PKCS12KeyStore.java]: Sure. Updated.

**Review by valeriepeng — APPROVED** [review]
Changes look good.

**Pull request closed without merging**